File: python/agentpin/bundle.py
"""Trust bundle support for AgentPin.

A trust bundle is a pre-shared collection of discovery and revocation
documents, enabling verification in environments where the standard
.well-known HTTP discovery is unavailable.
"""

import json
from datetime import datetime, timezone
from typing import Any, Dict, List, Optional

from .jwt import decode_jwt_unverified
from .pinning import KeyPinStore
from .types import ErrorCode, VerificationResult, VerifierConfig
from .verification import verify_credential_offline


def create_trust_bundle(created_at: Optional[str] = None) -> Dict[str, Any]:
    """Create a new empty trust bundle."""
    return {
        "agentpin_bundle_version": "0.1",
        "created_at": created_at or datetime.now(timezone.utc).isoformat(),
        "documents": [],
        "revocations": [],
    }


def find_bundle_discovery(bundle: Dict[str, Any], domain: str) -> Optional[Dict[str, Any]]:
    """Find a discovery document in a bundle by domain."""
    for doc in bundle.get("documents", []):
        if doc.get("entity") == domain:
            return doc
    return None


def find_bundle_revocation(bundle: Dict[str, Any], domain: str) -> Optional[Dict[str, Any]]:
    """Find a revocation document in a bundle by domain."""
    for doc in bundle.get("revocations", []):
        if doc.get("entity") == domain:
            return doc
    return None


def load_trust_bundle(path: str) -> Dict[str, Any]:
    """Load a trust bundle from a JSON file."""
    with open(path) as f:
        return json.load(f)


def save_trust_bundle(bundle: Dict[str, Any], path: str) -> None:
    """Save a trust bundle to a JSON file."""
    with open(path, "w") as f:
        json.dump(bundle, f, indent=2)


def verify_credential_with_bundle(
    credential_jwt: str,
    bundle: Dict[str, Any],
    pin_store: KeyPinStore,
    audience: Optional[str] = None,
    config: Optional[VerifierConfig] = None,
) -> VerificationResult:
    """Verify a credential using a trust bundle for discovery.

    Extracts the issuer domain from the JWT, looks up the discovery and
    revocation documents from the bundle, then delegates to
    verify_credential_offline.
    """
    if config is None:
        config = VerifierConfig()

    try:
        _header, payload, _sig = decode_jwt_unverified(credential_jwt)
    except Exception as e:
        return VerificationResult(
            valid=False,
            error_code=ErrorCode.ALGORITHM_REJECTED,
            error_message=f"JWT parse failed: {e}",
        )

    discovery = find_bundle_discovery(bundle, payload["iss"])
    if discovery is None:
        return VerificationResult(
            valid=False,
            error_code=ErrorCode.DISCOVERY_FETCH_FAILED,
            error_message=f"Domain '{payload['iss']}' not found in trust bundle",
        )

    revocation = find_bundle_revocation(bundle, payload["iss"])
    return verify_credential_offline(credential_jwt, discovery, revocation, pin_store, audience, config)

File: javascript/src/bundle.js
/**
 * Trust bundle support for AgentPin.
 *
 * A trust bundle is a pre-shared collection of discovery and revocation
 * documents, enabling verification in environments where the standard
 * .well-known HTTP discovery is unavailable.
 */

import { verifyCredentialOffline, defaultVerifierConfig } from './verification.js';
import { decodeJwtUnverified } from './jwt.js';
import { ErrorCode } from './types.js';

/**
 * Create a new empty trust bundle.
 * @param {string} createdAt - ISO 8601 timestamp
 * @returns {object}
 */
export function createTrustBundle(createdAt) {
    return {
        agentpin_bundle_version: '0.1',
        created_at: createdAt || new Date().toISOString(),
        documents: [],
        revocations: [],
    };
}

/**
 * Find a discovery document in a bundle by domain.
 * @param {object} bundle
 * @param {string} domain
 * @returns {object|null}
 */
export function findBundleDiscovery(bundle, domain) {
    return bundle.documents.find(d => d.entity === domain) || null;
}

/**
 * Find a revocation document in a bundle by domain.
 * @param {object} bundle
 * @param {string} domain
 * @returns {object|null}
 */
export function findBundleRevocation(bundle, domain) {
    return (bundle.revocations || []).find(r => r.entity === domain) || null;
}

/**
 * Verify a credential using a trust bundle for discovery.
 *
 * Extracts the issuer domain from the JWT, looks up the discovery and
 * revocation documents from the bundle, then delegates to
 * verifyCredentialOffline.
 *
 * @param {string} credentialJwt
 * @param {object} bundle - Trust bundle
 * @param {import('./pinning.js').KeyPinStore} pinStore
 * @param {string|null} audience
 * @param {object} [config]
 * @returns {object} VerificationResult
 */
export function verifyCredentialWithBundle(credentialJwt, bundle, pinStore, audience, config) {
    config = config || defaultVerifierConfig();

    let payload;
    try {
        ({ payload } = decodeJwtUnverified(credentialJwt));
    } catch (e) {
        return {
            valid: false,
            agent_id: null,
            issuer: null,
            capabilities: null,
            constraints: null,
            delegation_verified: null,
            delegation_chain: null,
            key_pinning: null,
            error_code: ErrorCode.ALGORITHM_REJECTED,
            error_message: `JWT parse failed: ${e.message}`,
            warnings: [],
        };
    }

    const discovery = findBundleDiscovery(bundle, payload.iss);
    if (!discovery) {
        return {
            valid: false,
            agent_id: null,
            issuer: null,
            capabilities: null,
            constraints: null,
            delegation_verified: null,
            delegation_chain: null,
            key_pinning: null,
            error_code: ErrorCode.DISCOVERY_FETCH_FAILED,
            error_message: `Domain '${payload.iss}' not found in trust bundle`,
            warnings: [],
        };
    }

    const revocation = findBundleRevocation(bundle, payload.iss);
    return verifyCredentialOffline(credentialJwt, discovery, revocation, pinStore, audience, config);
}
